### ISSUE TITLE
Errno 2 encountered when MaNGA ID run the first time

### ISSUE THREAD

**Issue by sjshamsi** [sdss/marvin#593]
Hello,

Part of my program gathers a specific list of MaNGA IDs and then runs methods like `mangaid2plateifu()` or initialises a Maps object with `Maps()` in a class. I find that I am not able to run these methods with a specific MaNGA ID the first time round, getting an `Errno 2` about a specific `.FITS` file. The message is as follows:
```
SDSS_ACCESS> syncing... please wait
SDSS_ACCESS> Done!
/homes/sshamsi/.local/lib/python3.5/site-packages/marvin/tools/maps.py:212: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/sdss_access/20190113_079/sdss_access_00.err' mode='w' encoding='UTF-8'>
  return super(Maps, self).download(path_type, **params)
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-47-d2ce71cb7216> in <module>
----> 1 ob = SpiralGalaxy(fits[106], 'emline_gflux_ha_6564')

<ipython-input-15-29a2dbced282> in __init__(self, file_path, emap_name)
      6         self.mangaid = self.data.metadata['MANGAID'][0]
      7         self.plateifu = mangaid2plateifu(self.mangaid, mode = 'auto')
----> 8         self.maps = Maps(self.plateifu, mode = 'auto')
      9         self.emap = self.maps[self.emap_name]
     10         self.is_usable = self.check_usability()

~/.local/lib/python3.5/site-packages/marvin/tools/maps.py in __init__(self, input, filename, mangaid, plateifu, mode, data, release, drpall, download, nsa_source, bintype, template, template_kin)
    106 
    107         if self.data_origin == 'file':
--> 108             self._load_maps_from_file(data=self.data)
    109         elif self.data_origin == 'db':
    110             self._load_maps_from_db(data=self.data)

~/.local/lib/python3.5/site-packages/marvin/tools/maps.py in _load_maps_from_file(self, data)
    241             assert isinstance(data, astropy.io.fits.HDUList), 'data is not a HDUList.'
    242         else:
--> 243             self.data = astropy.io.fits.open(self.filename)
    244 
    245         self.header = self.data[0].header

~/.local/lib/python3.5/site-packages/astropy/io/fits/hdu/hdulist.py in fitsopen(name, mode, memmap, save_backup, cache, lazy_load_hdus, **kwargs)
    149 
    150     return HDUList.fromfile(name, mode, memmap, save_backup, cache,
--> 151                             lazy_load_hdus, **kwargs)
    152 
    153 

~/.local/lib/python3.5/site-packages/astropy/io/fits/hdu/hdulist.py in fromfile(cls, fileobj, mode, memmap, save_backup, cache, lazy_load_hdus, **kwargs)
    388         return cls._readfrom(fileobj=fileobj, mode=mode, memmap=memmap,
    389                              save_backup=save_backup, cache=cache,
--> 390                              lazy_load_hdus=lazy_load_hdus, **kwargs)
    391 
    392     @classmethod

~/.local/lib/python3.5/site-packages/astropy/io/fits/hdu/hdulist.py in _readfrom(cls, fileobj, data, mode, memmap, save_backup, cache, lazy_load_hdus, **kwargs)
   1037             if not isinstance(fileobj, _File):
   1038                 # instantiate a FITS file object (ffo)
-> 1039                 fileobj = _File(fileobj, mode=mode, memmap=memmap, cache=cache)
   1040             # The Astropy mode is determined by the _File initializer if the
   1041             # supplied mode was None

~/.local/lib/python3.5/site-packages/astropy/utils/decorators.py in wrapper(*args, **kwargs)
    501                         # one with the name of the new argument to the function
    502                         kwargs[new_name[i]] = value
--> 503             return function(*args, **kwargs)
    504 
    505         return wrapper

~/.local/lib/python3.5/site-packages/astropy/io/fits/file.py in __init__(self, fileobj, mode, memmap, overwrite, cache)
    176             self._open_fileobj(fileobj, mode, overwrite)
    177         elif isinstance(fileobj, str):
--> 178             self._open_filename(fileobj, mode, overwrite)
    179         else:
    180             self._open_filelike(fileobj, mode, overwrite)

~/.local/lib/python3.5/site-packages/astropy/io/fits/file.py in _open_filename(self, filename, mode, overwrite)
    552         ext = os.path.splitext(self.name)[1]
    553 
--> 554         if not self._try_read_compressed(self.name, magic, mode, ext=ext):
    555             self._file = fileobj_open(self.name, IO_FITS_MODES[mode])
    556             self.close_on_error = True

~/.local/lib/python3.5/site-packages/astropy/io/fits/file.py in _try_read_compressed(self, obj_or_name, magic, mode, ext)
    450             else:
    451                 kwargs['fileobj'] = obj_or_name
--> 452             self._file = gzip.GzipFile(**kwargs)
    453             self.compression = 'gzip'
    454         elif ext == '.zip' or magic.startswith(PKZIP_MAGIC):

/usr/lib/python3.5/gzip.py in __init__(self, filename, mode, compresslevel, fileobj, mtime)
    161             mode += 'b'
    162         if fileobj is None:
--> 163             fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
    164         if filename is None:
    165             filename = getattr(fileobj, 'name', '')

FileNotFoundError: [Errno 2] No such file or directory: '/homes/sshamsi/sas/manga/spectro/analysis/v2_4_3/2.2.1/HYB10-GAU-MILESHC/8555/9101/manga-8555-9101-MAPS-HYB10-GAU-MILESHC.fits.gz'
```

However, when the same is run again, the methods are able to run and I am able to initialise my class objects.

**Desktop:**
 - Ubuntu
 - Chrome (running on Jupyterhub)
 - Marvin 2.3.0

**Comment by havok2063** [comment]
Hi @sjshamsi, it looks like you're trying to download the FITS files you need locally, is that correct?   I think this is related to how Marvin downloads the files it needs using `sdss_access`.  `sdss_access` downloads the files as a background task and nominally waits for the files to finish before moving on.  It sounds like your code or the Marvin code is continuing before the files have completed writing to disk, which causes the initial error.  When you rerun the code, it runs successfully since the files have finished writing to disk.  We've noticed this problem elsewhere.   We have a time delay added in, but we will try to fix this permanently.  You can try add a small time delay in your code as well.  

Alternatively, if you do not need the files locally, you can access the data remotely and you shouldn't have this problem.

**Comment by sjshamsi** [comment]
Hello @havok2063, one of the files my code depends on does download the .FITS files it needs. I tried adding the delay in my code in multiple locations but that does not seem to work for whatever reason. For now, I used a snipped where I just try to initialise my file again when the first error is raised, and it seems to be able to work because of that, albeit it's a bit slow.

When I try to use the remote mode, I come across the the issue as described in issue no. 594. Surprisingly, when I initialise my file the second time, I get a message:
```
SDSS_ACCESS> syncing... please wait
SDSS_ACCESS> Done!
```
Which makes me think that I am able to use `sdss_access`.
